File: scripts/solar/util/debug/DebugSgpUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 14,["7"] = 14,["8"] = 14,["10"] = 14,["11"] = 28,["12"] = 28,["13"] = 28,["15"] = 30,["16"] = 31,["17"] = 32,["18"] = 33,["22"] = 37,["23"] = 38,["24"] = 42,["25"] = 42,["27"] = 45,["28"] = 45,["29"] = 50,["30"] = 51,["31"] = 28,["32"] = 54,["33"] = 54,["34"] = 54,["36"] = 55,["37"] = 55,["38"] = 55,["39"] = 55,["40"] = 55,["41"] = 55,["42"] = 55,["43"] = 54,["44"] = 58,["45"] = 58,["46"] = 58,["48"] = 59,["49"] = 59,["50"] = 59,["51"] = 59,["52"] = 59,["53"] = 59,["54"] = 59,["55"] = 58,["56"] = 62,["57"] = 62,["58"] = 62,["60"] = 63,["61"] = 63,["62"] = 63,["63"] = 63,["64"] = 63,["65"] = 63,["66"] = 63,["67"] = 62,["68"] = 73,["69"] = 74,["70"] = 74,["71"] = 74,["72"] = 74,["73"] = 74,["74"] = 74,["75"] = 74,["76"] = 73,["77"] = 82,["78"] = 83,["79"] = 83,["80"] = 82,["81"] = 90,["82"] = 91,["83"] = 91,["84"] = 90,["85"] = 98,["86"] = 99,["87"] = 99,["88"] = 98,["89"] = 16,["90"] = 17,["91"] = 18,["92"] = 19,["93"] = 20,["94"] = 14});
local ____exports = {}
____exports.default = __TS__Class()
local DebugSgpUtil = ____exports.default
DebugSgpUtil.name = "DebugSgpUtil"
function DebugSgpUtil.prototype.____constructor(self)
end
function DebugSgpUtil.addCustomMenu(self, menuName, onAction, category, editorActionCommand, editorActionCommandParams)
    if category == nil then
        category = "基础"
    end
    local categoryMenu = nil
    for ____, temp in ipairs(____exports.default._sl_customMenus) do
        if temp.name == category then
            categoryMenu = temp
            break
        end
    end
    if not categoryMenu then
        categoryMenu = {name = category, menuItems = {}}
        local ____exports_default__sl_customMenus_0 = ____exports.default._sl_customMenus
        ____exports_default__sl_customMenus_0[#____exports_default__sl_customMenus_0 + 1] = categoryMenu
    end
    local ____categoryMenu_menuItems_1 = categoryMenu.menuItems
    ____categoryMenu_menuItems_1[#____categoryMenu_menuItems_1 + 1] = {name = menuName, editorActionCommand = editorActionCommand, editorActionCommandParams = editorActionCommandParams}
    local key = (category .. "_sl_") .. menuName
    ____exports.default._sl_customMenuActions[key] = onAction
end
function DebugSgpUtil.addCustomMenu_SelectUnitIds(self, menuName, onAction, category)
    if category == nil then
        category = "单位测试"
    end
    ____exports.default:addCustomMenu(
        menuName,
        onAction,
        category,
        "ObjectSelectPanel",
        "unit"
    )
end
function DebugSgpUtil.addCustomMenu_SelectItemIds(self, menuName, onAction, category)
    if category == nil then
        category = "物品测试"
    end
    ____exports.default:addCustomMenu(
        menuName,
        onAction,
        category,
        "ObjectSelectPanel",
        "unit"
    )
end
function DebugSgpUtil.addCustomMenu_SelectAbilityIds(self, menuName, onAction, category)
    if category == nil then
        category = "技能测试"
    end
    ____exports.default:addCustomMenu(
        menuName,
        onAction,
        category,
        "ObjectSelectPanel",
        "unit"
    )
end
function DebugSgpUtil.addCustomMenu_SelectCustomStrs(self, menuName, onAction, category, allOptionStrs)
    ____exports.default:addCustomMenu(
        menuName,
        onAction,
        category,
        "CustomStrsPanel",
        allOptionStrs
    )
end
function DebugSgpUtil.addOnCreateUnitListener(self, listener)
    local ____exports_default__sl_onCreateUnitListeners_2 = ____exports.default._sl_onCreateUnitListeners
    ____exports_default__sl_onCreateUnitListeners_2[#____exports_default__sl_onCreateUnitListeners_2 + 1] = listener
end
function DebugSgpUtil.addOnCreateItemListener(self, listener)
    local ____exports_default__sl_onCreateItemListeners_3 = ____exports.default._sl_onCreateItemListeners
    ____exports_default__sl_onCreateItemListeners_3[#____exports_default__sl_onCreateItemListeners_3 + 1] = listener
end
function DebugSgpUtil.addOnCreateAbilityUnitListener(self, listener)
    local ____exports_default__sl_onCreateAbilityUnitListeners_4 = ____exports.default._sl_onCreateAbilityUnitListeners
    ____exports_default__sl_onCreateAbilityUnitListeners_4[#____exports_default__sl_onCreateAbilityUnitListeners_4 + 1] = listener
end
DebugSgpUtil._sl_onCreateUnitListeners = {}
DebugSgpUtil._sl_onCreateItemListeners = {}
DebugSgpUtil._sl_onCreateAbilityUnitListeners = {}
DebugSgpUtil._sl_customMenus = {}
DebugSgpUtil._sl_customMenuActions = {}
____exports.default = DebugSgpUtil
return ____exports

File: src/solar/util/debug/DebugSgpUtil.ts
/**
 * sgp = solar game plugin （太阳编辑器游戏插件）
 * 可通过此类与左上角的太阳按钮交互 提交自定义的扩展功能 以方便调试
 */
type EditorActionCommand = "ObjectSelectPanel" | string
type CustomMenuType = {
    name: string//分类 一级菜单名
    menuItems: {
        name: string,//二级具体的可点击菜单的名字
        editorActionCommand?: EditorActionCommand //编辑器动作命令
        editorActionCommandParams?: any //编辑器动作命令的参数
    }[]
}
export default class DebugSgpUtil {

    static _sl_onCreateUnitListeners: ((unit: unit) => void)[] = [];
    static _sl_onCreateItemListeners: ((item: item) => void)[] = [];
    static _sl_onCreateAbilityUnitListeners: ((unit: unit, abilityIds: string[]) => void)[] = [];
    static _sl_customMenus: CustomMenuType[] = [];
    static _sl_customMenuActions: { [key: string]: (...args: any[]) => void } = {};


    /**
     * 添加自定义的菜单项
     * 请在地图游戏开始1.5秒内添加 之后添加的数据不会被刷新到太阳按钮上
     * editorActionCommand 会在点击onAction回调之前在编辑器里执行 并且将执行的结果传入onAction的参数列表
     */
    static addCustomMenu(menuName: string, onAction: (...args: any[]) => void, category: string = "基础",
                         editorActionCommand?: EditorActionCommand, editorActionCommandParams?: any) {
        let categoryMenu: CustomMenuType = null;
        for (let temp of DebugSgpUtil._sl_customMenus) {
            if (temp.name == category) {
                categoryMenu = temp;
                break
            }
        }
        if (!categoryMenu) {
            categoryMenu = {
                name: category,
                menuItems: []
            }
            DebugSgpUtil._sl_customMenus.push(categoryMenu);
        }
        //
        categoryMenu.menuItems.push({
            name: menuName,
            editorActionCommand: editorActionCommand,
            editorActionCommandParams: editorActionCommandParams
        });
        let key = category + "_sl_" + menuName;
        DebugSgpUtil._sl_customMenuActions[key] = onAction;
    }

    static addCustomMenu_SelectUnitIds(menuName: string, onAction: (unitIds: string[]) => void, category: string = "单位测试") {
        DebugSgpUtil.addCustomMenu(menuName, onAction, category, "ObjectSelectPanel", "unit")
    }

    static addCustomMenu_SelectItemIds(menuName: string, onAction: (itemIds: string[]) => void, category: string = "物品测试") {
        DebugSgpUtil.addCustomMenu(menuName, onAction, category, "ObjectSelectPanel", "unit")
    }

    static addCustomMenu_SelectAbilityIds(menuName: string, onAction: (abilityIds: string[]) => void, category: string = "技能测试") {
        DebugSgpUtil.addCustomMenu(menuName, onAction, category, "ObjectSelectPanel", "unit")
    }

    /**
     *
     * @param menuName 菜单名字
     * @param onAction 点击后的回调
     * @param category 菜单分类
     * @param allOptionStrs 所有字符串选项
     */
    static addCustomMenu_SelectCustomStrs(menuName: string, onAction: (choiceStrs: string[]) => void, category: string, allOptionStrs: string[]) {
        DebugSgpUtil.addCustomMenu(menuName, onAction, category, "CustomStrsPanel", allOptionStrs)
    }

    /**
     * 在左上角太阳按钮右键点击创建单位物编后会回调此方法
     * 添加创建单位的监听
     * @param listener
     */
    static addOnCreateUnitListener(listener: (unit: unit) => void) {
        DebugSgpUtil._sl_onCreateUnitListeners.push(listener)
    }

    /**
     * 创建物品物编时的回调
     * @param listener
     */
    static addOnCreateItemListener(listener: (item: item) => void) {
        DebugSgpUtil._sl_onCreateItemListeners.push(listener)
    }

    /**
     * 创建携带技能的单位
     * @param listener
     */
    static addOnCreateAbilityUnitListener(listener: (unit: unit, abilityIds: string[]) => void) {
        DebugSgpUtil._sl_onCreateAbilityUnitListeners.push(listener)
    }


}